modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/AbstractCppCodegen.java

```diff
@@ -353,4 +357,47 @@ public void preprocessOpenAPI(OpenAPI openAPI) {
         }
         return postProcessModelsEnum(objs);
     }
+
+    @Override
+    public Map<String, Object> postProcessAllModels(Map<String, Object> objs){
+        Map<String, Object> models = super.postProcessAllModels(objs);
+        for (final Entry<String, Object> model : models.entrySet()) {
+            CodegenModel mo = ModelUtils.getModelByName(model.getKey(), models);
+            addForwardDeclarations(mo, models);
+        }
+        return models;
+    }
+
+    private void addForwardDeclarations(CodegenModel model, Map<String, Object> objs) {
+        List<String> forwardDeclarations = new ArrayList<String>();
+        if(!model.hasVars) {
+            return;
+        }
+        for(CodegenProperty property : model.vars){
```

Comment:
I don't do much Java, but this loop looks odd, too much indentation levels IMO.

Half-assed attempt at fixing this a little - but I went down to 4 indentation levels.
I simply inverted the if statements, so they could probably be written in a smarter way.
```
for (CodegenProperty property : model.vars) {
	if (!(property.isContainer && property.mostInnerItems.isModel) && !(property.isModel))
		continue;
	
	String innerPropertyType = property.isContainer ? property.mostInnerItems.baseType : property.baseType;
	for (final Entry<String, Object> mo : objs.entrySet()) {
		CodegenModel innerModel = ModelUtils.getModelByName(mo.getKey(), objs);
		if (!(innerPropertyType.equals(innerModel.classname) && !innerPropertyType.equals(model.classname) && innerModel.hasVars))
			continue;
		for (CodegenProperty p : innerModel.vars) {
			if (!((p.isModel && p.dataType.equals(model.classname)) || (p.isContainer && p.mostInnerItems.baseType.equals(model.classname))))
				continue;
			String forwardDecl = "class " + innerModel.classname + ";";
			if (!forwardDeclarations.contains(forwardDecl)) {
				forwardDeclarations.add(forwardDecl);
			}
		}
	}
}
```

Reply:
I could make this readable like you suggested.
Need some feedback on the logic first too though.

Reply:
I refactored it a bit.